docs/memory-hotplug.md

```diff
@@ -0,0 +1,327 @@
+# Memory Hotplugging with virtio-mem
+
+## What is virtio-mem
+
+`virtio-mem` is a para-virtualized memory device that enables dynamic memory
+resizing for virtual machines. Unlike traditional memory hot-plug mechanisms,
+`virtio-mem` provides a flexible and efficient solution that works across
+different architectures and avoids many limitations of older approaches.
+
+The `virtio-mem` device manages a contiguous memory region that is divided into
+fixed-size blocks. The host can request the guest to plug (make available) or
+unplug (release) memory by changing the device's target size, and the guest
+driver responds by allocating or freeing memory blocks accordingly. This
+approach provides fine-grained control over guest memory with minimal overhead.
+
+Firecracker further adds the concept of slots, which are a set of contiguous
+blocks (usually 128MiB) that can be fully protected from guest accesses to
+prevent malicious guests from accessing the hotpluggable memory range when not
+allowed by the host.
+
+## Prerequisites
+
+To support memory hotplugging via `virtio-mem`, you must use a guest kernel with
+the appropriate version and configuration options enabled as follows:
+
+#### Kernel Version Requirements
+
+- `x86_64`: minimal kernel version is 5.16
+  - Firecracker requires support for `VIRTIO_MEM_F_UNPLUGGED_INACCESSIBLE`
+- `aarch64`: minimal kernel version is 5.18
+
+For more information about officially supported guest kernels, refer to the
+[kernel policy documentation](kernel-policy.md).
+
+#### Kernel Config
+
+The following kernel configuration options are required:
+
+```
+# Enable the virtio-mem driver
+CONFIG_VIRTIO_MEM=y
+
+# Enable memory hotplug support
+CONFIG_MEMORY_HOTPLUG=y
```

Comment:
just curious, if users enable just hotplug or hotremove, will be the feature functional for the corresponding counterpart?